content/blog/cran-checks-compiled-code/index.Rmd

```diff
@@ -0,0 +1,225 @@
+    char out[19 + 1];
+    int n;
+    n = snprintf(out, sizeof(out), "My height is %i cm", height);
+    Rprintf(out);
+    if (n >= sizeof(out)) {
+       Rprintf("\\nTruncated because input is longer than allowed!\\n");
+    }
+    return n;
+  }
+')
+
+say_height_safely(182)
+say_height_safely(1824)
+```
+
+If the destination is not a static buffer, the easiest thing to do is pass in the
+size of the array:
+
+```{r}
+cpp_function('
+  int say_height_safely(int height) {
+    // "My height is xxx cm" is 19 characters but we need 
+    // to add one for the null-terminator
+    size_t size = 19 + 1; 
+    char out[size]; 
+    int n; 
+    n = snprintf(out, size, "My height is %i cm", height);
+    Rprintf(out);
+    if (n >= sizeof(out)) {
+       Rprintf("\\nTruncated because input is longer than allowed!\\n");
+    }
+    return n;
+  }
+')
+
+say_height_safely(1824)
+```
+
+## WARNING regarding the use of strict prototypes in C
+
+Many maintainers with packages containing C code have also been getting hit with this warning:
+
+```
+warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
+```
+
+This usually comes from C function declarations that look like this, with no arguments specified (which is very common):
+
+``` c
+int myfun() {
+  ...
+};
+```
+
+This new warning is because CRAN is now running checks on R-devel with the `-Wstrict-prototypes` compiler flag set. In R we define functions that take no arguments with `myfun <- function() {...}` all the time. In C, with this flag set, the fact that a function takes no arguments must be explicitly stated (i.e., the arguments list cannot be empty). In the upcoming C23 standard, empty function signatures will be considered valid and not ambiguous, however at this point it is likely to be the reason you encounter this warning from CRAN.
+
+### How to fix it
+
+This can be fixed by placing the `void` keyword in the previously empty argument list:
+
+``` c
+int myfun(void) {
+  ...
+};
+```
+
+Here is an example where the authors of [Cubist](https://topepo.github.io/Cubist/) applied the [necessary patches](https://github.com/topepo/Cubist/pull/46), and [another one in rlang](https://github.com/r-lib/rlang/pull/1508).
+
+### Vendored code
+
+Function declarations without a prototype are very common, and unfortunately are thus likely to appear in libraries that you include in your package. This may require you to patch that code in your package. The [readxl](https://readxl.tidyverse.org) package includes the [libxls C library](https://github.com/libxls/libxls), which was patched [in readxl here](https://github.com/tidyverse/readxl/commit/afdc9b90cfc2bb1e1c5490c7ba3af5ecfc4a7876) to deal with this issue.
+
+The ideal solution in cases like this would be to submit patches to the upstream libraries so you don't have to deal with the ongoing maintenance of your local patches, but that is not always possible.
```

Comment:
Generally, however, you can explain this problem when submitting your package, and as long as you've have notified the upstream maintainer, CRAN will accept your updated package.